### ISSUE TITLE
listVirtualMachines does not show instances belonging to a project

### ISSUE THREAD

**Issue by NuxRo** [apache/cloudstack#3343]
Relevant email thread https://lists.apache.org/thread.html/6aca3e7b99816eef127ce516e5ef93bddb7a60a53e72762d5af3c050@%3Cdev.cloudstack.apache.org%3E

As a user I have created a project and in it I created a few VMs. These VMs will not show up in "listVirtualMachines" unless the project is specified.

e.g.
```
(localcloud) 🐱 > list virtualmachines
(localcloud) 🐱 > 
(localcloud) 🐱 > list virtualmachines projectid=b9a75ec4-82fc-4e88-8733-022df7956143 filter=id,
count = 2
virtualmachine:
id = 7b7259ac-ac89-43e7-9772-5f7866762be3
================================================================================
id = 0a1bb820-2ad6-4a71-bda0-29bb0ead6781
```

**Comment by NuxRo** [comment]
ping @ustcweizhou @PaulAngus

**Comment by PaulAngus** [comment]
@ustcweizhou did you say that you guys are planning on changing this behaviour or just that you would also like to see it changed?

**Comment by ustcweizhou** [comment]
@PaulAngus We have some changes in 4.7.1 to list all project resources in default view. That's the default behavior in our fork. As you can see, it breaks the backwards compatibility.
We need to make more changes before creating a pull request to mainstream. I will look into it.

**Comment by PaulAngus** [comment]
ok, cool.  I'll assign this to you if you don't mind so 'people' can see that it's kinda 'in-hand'.

**Comment by rhtyd** [comment]
@NuxRo known limitation, you've to pass projectid=-1 to list them all. This pattern exists for all list apis of various resources like routers, etc.

**Comment by ustcweizhou** [comment]
@rhtyd that's very nice. good to know it.
Can we add it cloudstack api document  ?
As far as I know, it exists for all commands extend BaseListProjectAndAccountResourcesCmd.java, right ?

**Comment by richardlawley** [comment]
Just come across this - would be nice if there was a way that you could list all "things" regardless of project or not.  Right now it looks like you have to do project=-1 to get anything in a project, and miss the project param for anything not in a project.

**Comment by rhtyd** [comment]
The workaround is to either specify a specific projectid=<uuid> or projectid=-1.

A resource under a project is owned by a project (a project account) than a member of the project, which is why such resources are not shown in the default list APIs/views. If we allow listing of all resources, then it may break the UI semantics (backward compatibility). A middle ground could be if `listall=true` is provided than all resources where user can access on could be listed on. Since we're close to 4.13.0.0, I've moved this to the next major milestone. cc @PaulAngus @ustcweizhou @NuxRo

**Comment by NuxRo** [comment]
Thanks Rohit, that makes sense. To be honest I would not have raised this issue had I known about projectid=-1, but that listall=true does not sound too bad if it could be done.

**Comment by weizhouapache** [comment]
probably fixed by #3894

**Comment by rhtyd** [comment]
Fixed in https://github.com/apache/cloudstack/pull/3894
